In [ ]:
###################

Student Name: Jing Zhu
Project: HAP725_W3W4_Assignment
    
##################

In [31]:
#Import libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import math

# Question 1. Construct a simple Multi-Morbidity Index

In [32]:
data = pd.read_excel('LM3_MM_Index.xlsx')

In [33]:
print(data)

    Case   MI  CHF   DM   AA   LOS    N
0      1   MI  CHF   DM  NaN  5.56   10
1      2   MI  NaN  NaN   AA  4.10   10
2      3  NaN  CHF   DM   AA  5.54   10
3      4  NaN  CHF   DM  NaN  3.56   20
4      5   MI  CHF  NaN   AA  7.03   30
5      6   MI  CHF  NaN  NaN  5.02   30
6      7  NaN  CHF  NaN   AA  5.04   30
7      8   MI  CHF   DM   AA  7.62   40
8      9   MI  NaN  NaN  NaN  2.03   40
9     10  NaN  CHF  NaN  NaN  3.03   40
10    11   MI  NaN   DM  NaN  2.60   50
11    12  NaN  NaN   DM   AA  2.57   50
12    13  NaN  NaN   DM  NaN  0.61   60
13    14  NaN  NaN  NaN   AA  2.12   70
14    15  NaN  NaN  NaN  NaN  0.01   80
15    16   MI  NaN   DM   AA  4.57  120


In [34]:
#Cover the diagnosises data into binary formats in order to be used to display the presense of each diagnosis.
data.loc[data['MI'] == 'MI', 'MIDX']=1
data.loc[data['MI'].isnull(), 'MIDX']=0
data.loc[data['CHF'] == 'CHF', 'CHFDX']=1
data.loc[data['CHF'].isnull(), 'CHFDX']=0
data.loc[data['DM'] == 'DM', 'DMDX']=1
data.loc[data['DM'].isnull(), 'DMDX']=0
data.loc[data['AA'] == 'AA', 'AADX']=1
data.loc[data['AA'].isnull(), 'AADX']=0

#View the dataframe head
data.head()

Case   MI  CHF   DM   AA   LOS   N  MIDX  CHFDX  DMDX  AADX
0     1   MI  CHF   DM  NaN  5.56  10   1.0    1.0   1.0   0.0
1     2   MI  NaN  NaN   AA  4.10  10   1.0    0.0   0.0   1.0
2     3  NaN  CHF   DM   AA  5.54  10   0.0    1.0   1.0   1.0
3     4  NaN  CHF   DM  NaN  3.56  20   0.0    1.0   1.0   0.0
4     5   MI  CHF  NaN   AA  7.03  30   1.0    1.0   0.0   1.0

In [35]:
#Put data ouput into dataframe
df=pd.DataFrame(data)

#Show the dataframe
df

Case   MI  CHF   DM   AA   LOS    N  MIDX  CHFDX  DMDX  AADX
0      1   MI  CHF   DM  NaN  5.56   10   1.0    1.0   1.0   0.0
1      2   MI  NaN  NaN   AA  4.10   10   1.0    0.0   0.0   1.0
2      3  NaN  CHF   DM   AA  5.54   10   0.0    1.0   1.0   1.0
3      4  NaN  CHF   DM  NaN  3.56   20   0.0    1.0   1.0   0.0
4      5   MI  CHF  NaN   AA  7.03   30   1.0    1.0   0.0   1.0
5      6   MI  CHF  NaN  NaN  5.02   30   1.0    1.0   0.0   0.0
6      7  NaN  CHF  NaN   AA  5.04   30   0.0    1.0   0.0   1.0
7      8   MI  CHF   DM   AA  7.62   40   1.0    1.0   1.0   1.0
8      9   MI  NaN  NaN  NaN  2.03   40   1.0    0.0   0.0   0.0
9     10  NaN  CHF  NaN  NaN  3.03   40   0.0    1.0   0.0   0.0
10    11   MI  NaN   DM  NaN  2.60   50   1.0    0.0   1.0   0.0
11    12  NaN  NaN   DM   AA  2.57   50   0.0    0.0   1.0   1.0
12    13  NaN  NaN   DM  NaN  0.61   60   0.0    0.0   1.0   0.0
13    14  NaN  NaN  NaN   AA  2.12   70   0.0    0.0   0.0   1.0
14    15  NaN  NaN  NaN  NaN  0.01   80   0.0    0.0   0.0   0.0
15    16   MI  NaN   DM   AA  4.57  120   1.0    0.0   1.0   1.0

In [36]:
#Calculate data values 
df['losMIDX'] = df['LOS']*df['MIDX']
df['losCHFDX'] = df['LOS']*df['CHFDX']
df['losDMDX'] = df['LOS']*df['DMDX']
df['losAADX'] = df['LOS']*df['AADX']

#Calculate Control values
df['losConMI']=(1-df['MIDX'])*df['LOS']
df['losConCHF']=(1-df['CHFDX'])*df['LOS']
df['losConDM']=(1-df['DMDX'])*df['LOS']
df['losConAA']=(1-df['AADX'])*df['LOS']

#Put data ouput into dataframe
df1 = pd.DataFrame(df)
df1

Case   MI  CHF   DM   AA   LOS    N  MIDX  CHFDX  DMDX  AADX  losMIDX  \
0      1   MI  CHF   DM  NaN  5.56   10   1.0    1.0   1.0   0.0     5.56   
1      2   MI  NaN  NaN   AA  4.10   10   1.0    0.0   0.0   1.0     4.10   
2      3  NaN  CHF   DM   AA  5.54   10   0.0    1.0   1.0   1.0     0.00   
3      4  NaN  CHF   DM  NaN  3.56   20   0.0    1.0   1.0   0.0     0.00   
4      5   MI  CHF  NaN   AA  7.03   30   1.0    1.0   0.0   1.0     7.03   
5      6   MI  CHF  NaN  NaN  5.02   30   1.0    1.0   0.0   0.0     5.02   
6      7  NaN  CHF  NaN   AA  5.04   30   0.0    1.0   0.0   1.0     0.00   
7      8   MI  CHF   DM   AA  7.62   40   1.0    1.0   1.0   1.0     7.62   
8      9   MI  NaN  NaN  NaN  2.03   40   1.0    0.0   0.0   0.0     2.03   
9     10  NaN  CHF  NaN  NaN  3.03   40   0.0    1.0   0.0   0.0     0.00   
10    11   MI  NaN   DM  NaN  2.60   50   1.0    0.0   1.0   0.0     2.60   
11    12  NaN  NaN   DM   AA  2.57   50   0.0    0.0   1.0   1.0     0.00   
12    13  NaN  NaN   DM  NaN  0.61   60   0.0    0.0   1.0   0.0     0.00   
13    14  NaN  NaN  NaN   AA  2.12   70   0.0    0.0   0.0   1.0     0.00   
14    15  NaN  NaN  NaN  NaN  0.01   80   0.0    0.0   0.0   0.0     0.00   
15    16   MI  NaN   DM   AA  4.57  120   1.0    0.0   1.0   1.0     4.57   

    losCHFDX  losDMDX  losAADX  losConMI  losConCHF  losConDM  losConAA  
0       5.56     5.56     0.00      0.00       0.00      0.00      5.56  
1       0.00     0.00     4.10      0.00       4.10      4.10      0.00  
2       5.54     5.54     5.54      5.54       0.00      0.00      0.00  
3       3.56     3.56     0.00      3.56       0.00      0.00      3.56  
4       7.03     0.00     7.03      0.00       0.00      7.03      0.00  
5       5.02     0.00     0.00      0.00       0.00      5.02      5.02  
6       5.04     0.00     5.04      5.04       0.00      5.04      0.00  
7       7.62     7.62     7.62      0.00       0.00      0.00      0.00  
8       0.00     0.00     0.00      0.00       2.03      2.03      2.03  
9       3.03     0.00     0.00      3.03       0.00      3.03      3.03  
10      0.00     2.60     0.00      0.00       2.60      0.00      2.60  
11      0.00     2.57     2.57      2.57       2.57      0.00      0.00  
12      0.00     0.61     0.00      0.61       0.61      0.00      0.61  
13      0.00     0.00     2.12      2.12       2.12      2.12      0.00  
14      0.00     0.00     0.00      0.01       0.01      0.01      0.01  
15      0.00     4.57     4.57      0.00       4.57      0.00      0.00

In [37]:
#Calculate the data value
df1['data_MI']=(df1['losMIDX'].sum()/df1['MIDX'].sum()).round(decimals=2)
df1['data_CHF']=(df1['losCHFDX'].sum()/df1['CHFDX'].sum()).round(decimals=2)
df1['data_DM']=(df1['losDMDX'].sum()/df1['DMDX'].sum()).round(decimals=2)
df1['data_AA']=(df1['losAADX'].sum()/df1['AADX'].sum()).round(decimals=2)

#Calculate the controls value
df1['Con_MI']=(df1['losConMI'].sum()/(1-df1['MIDX']).sum()).round(decimals=2)
df1['Con_CHF']=(df1['losConCHF'].sum()/(1-df1['CHFDX']).sum()).round(decimals=2)
df1['Con_DM']=(df1['losConDM'].sum()/(1-df1['DMDX']).sum()).round(decimals=2)
df1['Con_AA']=(df1['losConAA'].sum()/(1-df1['AADX']).sum()).round(decimals=2)

#Put data output into DataFrame
df2=pd.DataFrame(df1)
df2

Case   MI  CHF   DM   AA   LOS    N  MIDX  CHFDX  DMDX  ...  losConDM  \
0      1   MI  CHF   DM  NaN  5.56   10   1.0    1.0   1.0  ...      0.00   
1      2   MI  NaN  NaN   AA  4.10   10   1.0    0.0   0.0  ...      4.10   
2      3  NaN  CHF   DM   AA  5.54   10   0.0    1.0   1.0  ...      0.00   
3      4  NaN  CHF   DM  NaN  3.56   20   0.0    1.0   1.0  ...      0.00   
4      5   MI  CHF  NaN   AA  7.03   30   1.0    1.0   0.0  ...      7.03   
5      6   MI  CHF  NaN  NaN  5.02   30   1.0    1.0   0.0  ...      5.02   
6      7  NaN  CHF  NaN   AA  5.04   30   0.0    1.0   0.0  ...      5.04   
7      8   MI  CHF   DM   AA  7.62   40   1.0    1.0   1.0  ...      0.00   
8      9   MI  NaN  NaN  NaN  2.03   40   1.0    0.0   0.0  ...      2.03   
9     10  NaN  CHF  NaN  NaN  3.03   40   0.0    1.0   0.0  ...      3.03   
10    11   MI  NaN   DM  NaN  2.60   50   1.0    0.0   1.0  ...      0.00   
11    12  NaN  NaN   DM   AA  2.57   50   0.0    0.0   1.0  ...      0.00   
12    13  NaN  NaN   DM  NaN  0.61   60   0.0    0.0   1.0  ...      0.00   
13    14  NaN  NaN  NaN   AA  2.12   70   0.0    0.0   0.0  ...      2.12   
14    15  NaN  NaN  NaN  NaN  0.01   80   0.0    0.0   0.0  ...      0.01   
15    16   MI  NaN   DM   AA  4.57  120   1.0    0.0   1.0  ...      0.00   

    losConAA  data_MI  data_CHF  data_DM  data_AA  Con_MI  Con_CHF  Con_DM  \
0       5.56     4.82       5.3     4.08     4.82    2.81     2.33    3.55   
1       0.00     4.82       5.3     4.08     4.82    2.81     2.33    3.55   
2       0.00     4.82       5.3     4.08     4.82    2.81     2.33    3.55   
3       3.56     4.82       5.3     4.08     4.82    2.81     2.33    3.55   
4       0.00     4.82       5.3     4.08     4.82    2.81     2.33    3.55   
5       5.02     4.82       5.3     4.08     4.82    2.81     2.33    3.55   
6       0.00     4.82       5.3     4.08     4.82    2.81     2.33    3.55   
7       0.00     4.82       5.3     4.08     4.82    2.81     2.33    3.55   
8       2.03     4.82       5.3     4.08     4.82    2.81     2.33    3.55   
9       3.03     4.82       5.3     4.08     4.82    2.81     2.33    3.55   
10      2.60     4.82       5.3     4.08     4.82    2.81     2.33    3.55   
11      0.00     4.82       5.3     4.08     4.82    2.81     2.33    3.55   
12      0.61     4.82       5.3     4.08     4.82    2.81     2.33    3.55   
13      0.00     4.82       5.3     4.08     4.82    2.81     2.33    3.55   
14      0.01     4.82       5.3     4.08     4.82    2.81     2.33    3.55   
15      0.00     4.82       5.3     4.08     4.82    2.81     2.33    3.55   

    Con_AA  
0      2.8  
1      2.8  
2      2.8  
3      2.8  
4      2.8  
5      2.8  
6      2.8  
7      2.8  
8      2.8  
9      2.8  
10     2.8  
11     2.8  
12     2.8  
13     2.8  
14     2.8  
15     2.8  

[16 rows x 27 columns]

In [38]:
###To calculate the estimate
#Estimate = LOS for cases - LOS Control

#Calculate Estimate Values
df2['Est_MI']=df2['data_MI']-df2['Con_MI']
df2['Est_CHF']=df2['data_CHF']-df2['Con_CHF']
df2['Est_DM']=df2['data_DM']-df2['Con_DM']
df2['Est_AA']=df2['data_AA']-df2['Con_AA']

#Put data output into dataframe
df3 = pd.DataFrame(df2)
df3

Case   MI  CHF   DM   AA   LOS    N  MIDX  CHFDX  DMDX  ...  data_DM  \
0      1   MI  CHF   DM  NaN  5.56   10   1.0    1.0   1.0  ...     4.08   
1      2   MI  NaN  NaN   AA  4.10   10   1.0    0.0   0.0  ...     4.08   
2      3  NaN  CHF   DM   AA  5.54   10   0.0    1.0   1.0  ...     4.08   
3      4  NaN  CHF   DM  NaN  3.56   20   0.0    1.0   1.0  ...     4.08   
4      5   MI  CHF  NaN   AA  7.03   30   1.0    1.0   0.0  ...     4.08   
5      6   MI  CHF  NaN  NaN  5.02   30   1.0    1.0   0.0  ...     4.08   
6      7  NaN  CHF  NaN   AA  5.04   30   0.0    1.0   0.0  ...     4.08   
7      8   MI  CHF   DM   AA  7.62   40   1.0    1.0   1.0  ...     4.08   
8      9   MI  NaN  NaN  NaN  2.03   40   1.0    0.0   0.0  ...     4.08   
9     10  NaN  CHF  NaN  NaN  3.03   40   0.0    1.0   0.0  ...     4.08   
10    11   MI  NaN   DM  NaN  2.60   50   1.0    0.0   1.0  ...     4.08   
11    12  NaN  NaN   DM   AA  2.57   50   0.0    0.0   1.0  ...     4.08   
12    13  NaN  NaN   DM  NaN  0.61   60   0.0    0.0   1.0  ...     4.08   
13    14  NaN  NaN  NaN   AA  2.12   70   0.0    0.0   0.0  ...     4.08   
14    15  NaN  NaN  NaN  NaN  0.01   80   0.0    0.0   0.0  ...     4.08   
15    16   MI  NaN   DM   AA  4.57  120   1.0    0.0   1.0  ...     4.08   

    data_AA  Con_MI  Con_CHF  Con_DM  Con_AA  Est_MI  Est_CHF  Est_DM  Est_AA  
0      4.82    2.81     2.33    3.55     2.8    2.01     2.97    0.53    2.02  
1      4.82    2.81     2.33    3.55     2.8    2.01     2.97    0.53    2.02  
2      4.82    2.81     2.33    3.55     2.8    2.01     2.97    0.53    2.02  
3      4.82    2.81     2.33    3.55     2.8    2.01     2.97    0.53    2.02  
4      4.82    2.81     2.33    3.55     2.8    2.01     2.97    0.53    2.02  
5      4.82    2.81     2.33    3.55     2.8    2.01     2.97    0.53    2.02  
6      4.82    2.81     2.33    3.55     2.8    2.01     2.97    0.53    2.02  
7      4.82    2.81     2.33    3.55     2.8    2.01     2.97    0.53    2.02  
8      4.82    2.81     2.33    3.55     2.8    2.01     2.97    0.53    2.02  
9      4.82    2.81     2.33    3.55     2.8    2.01     2.97    0.53    2.02  
10     4.82    2.81     2.33    3.55     2.8    2.01     2.97    0.53    2.02  
11     4.82    2.81     2.33    3.55     2.8    2.01     2.97    0.53    2.02  
12     4.82    2.81     2.33    3.55     2.8    2.01     2.97    0.53    2.02  
13     4.82    2.81     2.33    3.55     2.8    2.01     2.97    0.53    2.02  
14     4.82    2.81     2.33    3.55     2.8    2.01     2.97    0.53    2.02  
15     4.82    2.81     2.33    3.55     2.8    2.01     2.97    0.53    2.02  

[16 rows x 31 columns]

In [39]:
#Caculate Severity
df3['severity']=((df3['MIDX']*df3['Est_MI'])+(df3['CHFDX']*df3['Est_CHF']
                 +(df3['DMDX']*df3['Est_DM'])+(df3['AADX']*df3['Est_AA'])).round(decimals=2))

#Put data output into dataframe
df4 = pd.DataFrame(df3)
df4

Case   MI  CHF   DM   AA   LOS    N  MIDX  CHFDX  DMDX  ...  data_AA  \
0      1   MI  CHF   DM  NaN  5.56   10   1.0    1.0   1.0  ...     4.82   
1      2   MI  NaN  NaN   AA  4.10   10   1.0    0.0   0.0  ...     4.82   
2      3  NaN  CHF   DM   AA  5.54   10   0.0    1.0   1.0  ...     4.82   
3      4  NaN  CHF   DM  NaN  3.56   20   0.0    1.0   1.0  ...     4.82   
4      5   MI  CHF  NaN   AA  7.03   30   1.0    1.0   0.0  ...     4.82   
5      6   MI  CHF  NaN  NaN  5.02   30   1.0    1.0   0.0  ...     4.82   
6      7  NaN  CHF  NaN   AA  5.04   30   0.0    1.0   0.0  ...     4.82   
7      8   MI  CHF   DM   AA  7.62   40   1.0    1.0   1.0  ...     4.82   
8      9   MI  NaN  NaN  NaN  2.03   40   1.0    0.0   0.0  ...     4.82   
9     10  NaN  CHF  NaN  NaN  3.03   40   0.0    1.0   0.0  ...     4.82   
10    11   MI  NaN   DM  NaN  2.60   50   1.0    0.0   1.0  ...     4.82   
11    12  NaN  NaN   DM   AA  2.57   50   0.0    0.0   1.0  ...     4.82   
12    13  NaN  NaN   DM  NaN  0.61   60   0.0    0.0   1.0  ...     4.82   
13    14  NaN  NaN  NaN   AA  2.12   70   0.0    0.0   0.0  ...     4.82   
14    15  NaN  NaN  NaN  NaN  0.01   80   0.0    0.0   0.0  ...     4.82   
15    16   MI  NaN   DM   AA  4.57  120   1.0    0.0   1.0  ...     4.82   

    Con_MI  Con_CHF  Con_DM  Con_AA  Est_MI  Est_CHF  Est_DM  Est_AA  severity  
0     2.81     2.33    3.55     2.8    2.01     2.97    0.53    2.02      5.51  
1     2.81     2.33    3.55     2.8    2.01     2.97    0.53    2.02      4.03  
2     2.81     2.33    3.55     2.8    2.01     2.97    0.53    2.02      5.52  
3     2.81     2.33    3.55     2.8    2.01     2.97    0.53    2.02      3.50  
4     2.81     2.33    3.55     2.8    2.01     2.97    0.53    2.02      7.00  
5     2.81     2.33    3.55     2.8    2.01     2.97    0.53    2.02      4.98  
6     2.81     2.33    3.55     2.8    2.01     2.97    0.53    2.02      4.99  
7     2.81     2.33    3.55     2.8    2.01     2.97    0.53    2.02      7.53  
8     2.81     2.33    3.55     2.8    2.01     2.97    0.53    2.02      2.01  
9     2.81     2.33    3.55     2.8    2.01     2.97    0.53    2.02      2.97  
10    2.81     2.33    3.55     2.8    2.01     2.97    0.53    2.02      2.54  
11    2.81     2.33    3.55     2.8    2.01     2.97    0.53    2.02      2.55  
12    2.81     2.33    3.55     2.8    2.01     2.97    0.53    2.02      0.53  
13    2.81     2.33    3.55     2.8    2.01     2.97    0.53    2.02      2.02  
14    2.81     2.33    3.55     2.8    2.01     2.97    0.53    2.02      0.00  
15    2.81     2.33    3.55     2.8    2.01     2.97    0.53    2.02      4.56  

[16 rows x 32 columns]

# Question 2. Calculate the likelihood ratio associated

In [40]:
data = pd.read_excel('LM3_MM_Index.xlsx')

In [41]:
print(data)

    Case   MI  CHF   DM   AA   LOS    N
0      1   MI  CHF   DM  NaN  5.56   10
1      2   MI  NaN  NaN   AA  4.10   10
2      3  NaN  CHF   DM   AA  5.54   10
3      4  NaN  CHF   DM  NaN  3.56   20
4      5   MI  CHF  NaN   AA  7.03   30
5      6   MI  CHF  NaN  NaN  5.02   30
6      7  NaN  CHF  NaN   AA  5.04   30
7      8   MI  CHF   DM   AA  7.62   40
8      9   MI  NaN  NaN  NaN  2.03   40
9     10  NaN  CHF  NaN  NaN  3.03   40
10    11   MI  NaN   DM  NaN  2.60   50
11    12  NaN  NaN   DM   AA  2.57   50
12    13  NaN  NaN   DM  NaN  0.61   60
13    14  NaN  NaN  NaN   AA  2.12   70
14    15  NaN  NaN  NaN  NaN  0.01   80
15    16   MI  NaN   DM   AA  4.57  120


In [42]:
data['N']=data['N'].astype(float)
print(data)

    Case   MI  CHF   DM   AA   LOS      N
0      1   MI  CHF   DM  NaN  5.56   10.0
1      2   MI  NaN  NaN   AA  4.10   10.0
2      3  NaN  CHF   DM   AA  5.54   10.0
3      4  NaN  CHF   DM  NaN  3.56   20.0
4      5   MI  CHF  NaN   AA  7.03   30.0
5      6   MI  CHF  NaN  NaN  5.02   30.0
6      7  NaN  CHF  NaN   AA  5.04   30.0
7      8   MI  CHF   DM   AA  7.62   40.0
8      9   MI  NaN  NaN  NaN  2.03   40.0
9     10  NaN  CHF  NaN  NaN  3.03   40.0
10    11   MI  NaN   DM  NaN  2.60   50.0
11    12  NaN  NaN   DM   AA  2.57   50.0
12    13  NaN  NaN   DM  NaN  0.61   60.0
13    14  NaN  NaN  NaN   AA  2.12   70.0
14    15  NaN  NaN  NaN  NaN  0.01   80.0
15    16   MI  NaN   DM   AA  4.57  120.0


In [43]:
AvgLos = (data["LOS"]*data["N"]).sum()/data["N"].sum()
print(AvgLos)

3.240289855072464


In [44]:
print(data)

    Case   MI  CHF   DM   AA   LOS      N
0      1   MI  CHF   DM  NaN  5.56   10.0
1      2   MI  NaN  NaN   AA  4.10   10.0
2      3  NaN  CHF   DM   AA  5.54   10.0
3      4  NaN  CHF   DM  NaN  3.56   20.0
4      5   MI  CHF  NaN   AA  7.03   30.0
5      6   MI  CHF  NaN  NaN  5.02   30.0
6      7  NaN  CHF  NaN   AA  5.04   30.0
7      8   MI  CHF   DM   AA  7.62   40.0
8      9   MI  NaN  NaN  NaN  2.03   40.0
9     10  NaN  CHF  NaN  NaN  3.03   40.0
10    11   MI  NaN   DM  NaN  2.60   50.0
11    12  NaN  NaN   DM   AA  2.57   50.0
12    13  NaN  NaN   DM  NaN  0.61   60.0
13    14  NaN  NaN  NaN   AA  2.12   70.0
14    15  NaN  NaN  NaN  NaN  0.01   80.0
15    16   MI  NaN   DM   AA  4.57  120.0


In [45]:
data.replace(to_replace = ['MI','CHF','DM','AA'],
              value = ['1','1','1','1'],
              inplace = True)
data = data.fillna(0)
data["MI"]=data["MI"].astype(float)
data["CHF"]=data["CHF"].astype(float)
data["DM"]=data["DM"].astype(float)
data["AA"]=data["AA"].astype(float)


In [46]:
data['TempMILong'] = np.where(data['LOS']>AvgLos,'1','0')
data["TempMILong"] = data["TempMILong"].astype(float)
data['TempMIShort'] = np.where(data['LOS']<=AvgLos,'1','0')
data["TempMIShort"] = data["TempMIShort"].astype(float)
data

Case   MI  CHF   DM   AA   LOS      N  TempMILong  TempMIShort
0      1  1.0  1.0  1.0  0.0  5.56   10.0         1.0          0.0
1      2  1.0  0.0  0.0  1.0  4.10   10.0         1.0          0.0
2      3  0.0  1.0  1.0  1.0  5.54   10.0         1.0          0.0
3      4  0.0  1.0  1.0  0.0  3.56   20.0         1.0          0.0
4      5  1.0  1.0  0.0  1.0  7.03   30.0         1.0          0.0
5      6  1.0  1.0  0.0  0.0  5.02   30.0         1.0          0.0
6      7  0.0  1.0  0.0  1.0  5.04   30.0         1.0          0.0
7      8  1.0  1.0  1.0  1.0  7.62   40.0         1.0          0.0
8      9  1.0  0.0  0.0  0.0  2.03   40.0         0.0          1.0
9     10  0.0  1.0  0.0  0.0  3.03   40.0         0.0          1.0
10    11  1.0  0.0  1.0  0.0  2.60   50.0         0.0          1.0
11    12  0.0  0.0  1.0  1.0  2.57   50.0         0.0          1.0
12    13  0.0  0.0  1.0  0.0  0.61   60.0         0.0          1.0
13    14  0.0  0.0  0.0  1.0  2.12   70.0         0.0          1.0
14    15  0.0  0.0  0.0  0.0  0.01   80.0         0.0          1.0
15    16  1.0  0.0  1.0  1.0  4.57  120.0         1.0          0.0

In [49]:
data['TempAALong'] = np.where(data['LOS']>AvgLos,'1','0')
data["TempAALong"] = data["TempAALong"].astype(float)
data['TempAAShort'] = np.where(data['LOS']<=AvgLos,'1','0')
data["TempAAShort"] = data["TempAAShort"].astype(float)
data

Case   MI  CHF   DM   AA   LOS      N  TempMILong  TempMIShort TempALong  \
0      1  1.0  1.0  1.0  0.0  5.56   10.0         1.0          0.0         1   
1      2  1.0  0.0  0.0  1.0  4.10   10.0         1.0          0.0         1   
2      3  0.0  1.0  1.0  1.0  5.54   10.0         1.0          0.0         1   
3      4  0.0  1.0  1.0  0.0  3.56   20.0         1.0          0.0         1   
4      5  1.0  1.0  0.0  1.0  7.03   30.0         1.0          0.0         1   
5      6  1.0  1.0  0.0  0.0  5.02   30.0         1.0          0.0         1   
6      7  0.0  1.0  0.0  1.0  5.04   30.0         1.0          0.0         1   
7      8  1.0  1.0  1.0  1.0  7.62   40.0         1.0          0.0         1   
8      9  1.0  0.0  0.0  0.0  2.03   40.0         0.0          1.0         0   
9     10  0.0  1.0  0.0  0.0  3.03   40.0         0.0          1.0         0   
10    11  1.0  0.0  1.0  0.0  2.60   50.0         0.0          1.0         0   
11    12  0.0  0.0  1.0  1.0  2.57   50.0         0.0          1.0         0   
12    13  0.0  0.0  1.0  0.0  0.61   60.0         0.0          1.0         0   
13    14  0.0  0.0  0.0  1.0  2.12   70.0         0.0          1.0         0   
14    15  0.0  0.0  0.0  0.0  0.01   80.0         0.0          1.0         0   
15    16  1.0  0.0  1.0  1.0  4.57  120.0         1.0          0.0         1   

    TempAALong  TempAAShort  
0          1.0          0.0  
1          1.0          0.0  
2          1.0          0.0  
3          1.0          0.0  
4          1.0          0.0  
5          1.0          0.0  
6          1.0          0.0  
7          1.0          0.0  
8          0.0          1.0  
9          0.0          1.0  
10         0.0          1.0  
11         0.0          1.0  
12         0.0          1.0  
13         0.0          1.0  
14         0.0          1.0  
15         1.0          0.0

In [53]:
MILong=(data['MI']*data['N']*data["TempMILong"]).sum()
print(MILong)
MIShort=(data['MI']*data['N']*data["TempMIShort"]).sum()
print(MIShort)

AALong=(data['AA']*data['N']*data["TempAALong"]).sum()
print(AALong)
AAShort=(data['AA']*data['N']*data["TempAAShort"]).sum()
print(AAShort)

240.0
90.0
240.0
120.0


In [54]:
LongLOS=(data['N']*data["TempMILong"]).sum()
print(LongLOS)
ShortLOS=(data['N']*data["TempMIShort"]).sum()
print(ShortLOS)
AALongLOS=(data['N']*data["TempAALong"]).sum()
print(AALongLOS)
AAShortLOS=(data['N']*data["TempAAShort"]).sum()
print(AAShortLOS)

300.0
390.0
300.0
390.0


In [56]:
MILIkelihoodRatio = [(MILong/LongLOS)/(MIShort/ShortLOS)]
print(MILIkelihoodRatio)
AALIkelihoodRatio = [(AALong/AALongLOS)/(AAShort/AAShortLOS)]
print(AALIkelihoodRatio)

[3.466666666666667]
[2.6]


Conclusion: Patients who have MI are 3.47 more likely to stay longer at the hospital.

# Question 3. Calculate the probability of long stay for the 3 patients 

In [25]:
data = pd.read_excel('W4Q3Data.xlsx')

In [24]:
data = pd.DataFrame(data, columns = ['ID', 'ICD9', 'LR'])
print(data)

   ID ICD9    LR
0   1   AA  2.60
1   1   DM  1.63
2   1  HIV  1.00
3   2   MI  3.47
4   2  CHF  5.53
5   3   DM  1.63
6   3  CHF  5.53
7   3   AA  2.60
8   3   MI  3.47


In [26]:
#axis=0 does it by row:axis=1 does it by column
pd = data.filter(items=[0,1,2],axis=0)
pd

ID ICD9    LR
0   1   AA  2.60
1   1   DM  1.63
2   1  HIV  1.00

In [28]:
###Calculate probability of LR of long stay
#Take natural Log of LR
pd['log']=np.log(data['LR'])
pd

ID ICD9    LR       log
0   1   AA  2.60  0.955511
1   1   DM  1.63  0.488580
2   1  HIV  1.00  0.000000

In [29]:
###Calculate odds
pd['Odds']=np.sum(pd['log'])
pd['Odds']=pd['Odds'].astype(float)
pd['Odds']=np.exp(pd['Odds'])
pd

ID ICD9    LR       log   Odds
0   1   AA  2.60  0.955511  4.238
1   1   DM  1.63  0.488580  4.238
2   1  HIV  1.00  0.000000  4.238

In [32]:
###Probability
pd['Prob']=(pd['Odds']/(pd['Odds']+1))
pd

ID ICD9    LR       log   Odds      Prob
0   1   AA  2.60  0.955511  4.238  0.809087
1   1   DM  1.63  0.488580  4.238  0.809087
2   1  HIV  1.00  0.000000  4.238  0.809087

In [36]:
#id2
pd=data.filter(items=[3,4],axis=0)

#Log and oddspd
pd['log']=np.log(data['LR'])
pd['Odds']=np.sum(pd['log'])
pd['Odds']=pd['Odds'].astype(float)
pd['Odds']=np.exp(pd['Odds'])
pd

#Probabilty
pd['Prob']=(pd['Odds']/(pd['Odds']+1))
pd

#id3
pd1=data.filter(items=[5,6,7,8],axis=0)
pd

ID ICD9    LR       log     Odds      Prob
3   2   MI  3.47  1.244155  19.1891  0.950468
4   2  CHF  5.53  1.710188  19.1891  0.950468

In [37]:
#Log
pd1['log']=np.log(data['LR'])

#Odds
pd1['Odds']=np.sum(pd1['log'])
pd1['Odds']=pd1['Odds'].astype(float)
pd1['Odds']=np.exp(pd1['Odds'])

#probability
pd1['Prob']=(pd1['Odds']/(pd1['Odds']+1))
pd1

ID ICD9    LR       log       Odds      Prob
5   3   DM  1.63  0.488580  81.323406  0.987853
6   3  CHF  5.53  1.710188  81.323406  0.987853
7   3   AA  2.60  0.955511  81.323406  0.987853
8   3   MI  3.47  1.244155  81.323406  0.987853

# Question 7

In [2]:
data = pd.read_csv('W4Q7HW.csv')

In [4]:
data.info

<bound method DataFrame.info of         id     icd9  rdxcalc  RDX        LR  nDead   nAlive  TotalDead  \
0    19996  IV62.84        1  1.0  0.063713  752.0  59400.0   121567.0   
1    19981  IV62.84        3  3.0  0.068872  124.0   9061.0   121567.0   
2    19981  IV62.85        1  1.0  0.046855   88.0   9452.0   121567.0   
3    19981   IV65.2        1  1.0  0.035061   24.0   3445.0   121567.0   
4    19988  IV85.36        1  1.0  0.105977   43.0   2042.0   121567.0   
..     ...      ...      ...  ...       ...    ...      ...        ...   
256  19991   IV60.2        1  1.0  0.033854   91.0  13528.0   121567.0   
257  19981   IV60.9        1  1.0  0.056938   18.0   1591.0   121567.0   
258  19985   IV62.0        1  1.0  0.027342  197.0  36260.0   121567.0   
259  19991   IV62.0        1  1.0  0.027342  197.0  36260.0   121567.0   
260  19996   I585.9        6  NaN       NaN    NaN      NaN        NaN   

     TotalNotDead  
0        611807.0  
1        611807.0  
2        611807.0  

In [5]:
##### Finding Log####
data['log']=np.log10(data["LR"])
data.head()

id     icd9  rdxcalc  RDX        LR  nDead   nAlive  TotalDead  \
0  19996  IV62.84        1  1.0  0.063713  752.0  59400.0   121567.0   
1  19981  IV62.84        3  3.0  0.068872  124.0   9061.0   121567.0   
2  19981  IV62.85        1  1.0  0.046855   88.0   9452.0   121567.0   
3  19981   IV65.2        1  1.0  0.035061   24.0   3445.0   121567.0   
4  19988  IV85.36        1  1.0  0.105977   43.0   2042.0   121567.0   

   TotalNotDead       log  
0      611807.0 -1.195770  
1      611807.0 -1.161956  
2      611807.0 -1.329242  
3      611807.0 -1.455179  
4      611807.0 -0.974789

In [6]:
#### Finding the sum of log####
result = data.groupby("id")["log"].sum()

###Put result output data into dataframe###
result_log=pd.DataFrame(result)
result_log

log
id              
19980  -2.226542
19981 -19.119586
19982  -3.685418
19983  -3.709157
19984  -4.891829
19985  -7.030168
19986  -2.768233
19987  -4.829137
19988 -10.998985
19989  -3.486235
19990  -5.403624
19991 -13.700944
19992  -2.234513
19993  -0.315481
19994  -3.895245
19995  -1.750001
19996 -21.492348
19997  -9.817192
19998  -6.500803
19999 -10.981587
20000  -5.705737

In [8]:
#### Calculating Odds ####
result_log["Odds"]=np.exp(result_log["log"])
result_Odds=pd.DataFrame(result_log)
result_Odds

log          Odds
id                            
19980  -2.226542  1.079009e-01
19981 -19.119586  4.971293e-09
19982  -3.685418  2.508669e-02
19983  -3.709157  2.449817e-02
19984  -4.891829  7.507678e-03
19985  -7.030168  8.847835e-04
19986  -2.768233  6.277284e-02
19987  -4.829137  7.993416e-03
19988 -10.998985  1.671866e-05
19989  -3.486235  3.061593e-02
19990  -5.403624  4.500241e-03
19991 -13.700944  1.121387e-06
19992  -2.234513  1.070443e-01
19993  -0.315481  7.294376e-01
19994  -3.895245  2.033838e-02
19995  -1.750001  1.737738e-01
19996 -21.492348  4.634384e-10
19997  -9.817192  5.450645e-05
19998  -6.500803  1.502232e-03
19999 -10.981587  1.701207e-05
20000  -5.705737  3.326824e-03

In [9]:
result_Odds["Probability"]=((result_Odds["Odds"])/(1+(result_Odds["Odds"]))).round(5)
result_Prob=pd.DataFrame(result_Odds)
result_Prob

log          Odds  Probability
id                                         
19980  -2.226542  1.079009e-01      0.09739
19981 -19.119586  4.971293e-09      0.00000
19982  -3.685418  2.508669e-02      0.02447
19983  -3.709157  2.449817e-02      0.02391
19984  -4.891829  7.507678e-03      0.00745
19985  -7.030168  8.847835e-04      0.00088
19986  -2.768233  6.277284e-02      0.05907
19987  -4.829137  7.993416e-03      0.00793
19988 -10.998985  1.671866e-05      0.00002
19989  -3.486235  3.061593e-02      0.02971
19990  -5.403624  4.500241e-03      0.00448
19991 -13.700944  1.121387e-06      0.00000
19992  -2.234513  1.070443e-01      0.09669
19993  -0.315481  7.294376e-01      0.42178
19994  -3.895245  2.033838e-02      0.01993
19995  -1.750001  1.737738e-01      0.14805
19996 -21.492348  4.634384e-10      0.00000
19997  -9.817192  5.450645e-05      0.00005
19998  -6.500803  1.502232e-03      0.00150
19999 -10.981587  1.701207e-05      0.00002
20000  -5.705737  3.326824e-03      0.00332

In [10]:
### sort values by Probability Descending ###
result_Prob.sort_values(by="Probability",ascending=False)

log          Odds  Probability
id                                         
19993  -0.315481  7.294376e-01      0.42178
19995  -1.750001  1.737738e-01      0.14805
19980  -2.226542  1.079009e-01      0.09739
19992  -2.234513  1.070443e-01      0.09669
19986  -2.768233  6.277284e-02      0.05907
19989  -3.486235  3.061593e-02      0.02971
19982  -3.685418  2.508669e-02      0.02447
19983  -3.709157  2.449817e-02      0.02391
19994  -3.895245  2.033838e-02      0.01993
19987  -4.829137  7.993416e-03      0.00793
19984  -4.891829  7.507678e-03      0.00745
19990  -5.403624  4.500241e-03      0.00448
20000  -5.705737  3.326824e-03      0.00332
19998  -6.500803  1.502232e-03      0.00150
19985  -7.030168  8.847835e-04      0.00088
19997  -9.817192  5.450645e-05      0.00005
19988 -10.998985  1.671866e-05      0.00002
19999 -10.981587  1.701207e-05      0.00002
19991 -13.700944  1.121387e-06      0.00000
19981 -19.119586  4.971293e-09      0.00000
19996 -21.492348  4.634384e-10      0.00000